README.md
# Pirate Joe

[![Build Status][travis-image]][travis-url]
[![Coverage Status][coveralls-image]][coveralls-url]
[![js-standard-style][standard-image]][standard-url]
[![standard-version][standard-version-image]][standard-version-url]

Yargs' slack-bot Pirate Joe, the eternal optimist.

![Joe was one optimistic pirate.](http://i.imgur.com/4WFGVJ9.png)

## Building Slack a Slack-Bot with yargs

yargs, with its new [headless `parse()`](https://github.com/yargs/yargs#parseargs-context-parsecallback) feature and its powerful command
parsing functionality, is a great tool for building a chat-bot.

The [Pirate Joe sample project](https://github.com/yargs/pirate-joe/blob/master/server.js) and accompanying tutorial provide the foundation you need to get a chat-bot up in running in Slack using yargs.

### Parsing Incoming Messages

We've implemented Pirate Joe as a [Slack slash-command](https://api.slack.com/slash-commands). When
a user types `/joe` in any of yargs' slack channels, a webhook is posted to whatever URL you configure.

To process this webhook we created a [minimal express application](https://expressjs.com/en/starter/hello-world.html), with the the following `post` handler:

```js
const bodyParser = require('body-parser')
const express = require('express')

// Slack's slash commands are passed as an urlencoded
// HTTP post: https://api.slack.com/slash-commands
app.use(bodyParser.urlencoded({ extended: false }))

// slack webhook endpoint.
app.post('/', function (req, res) {
  let context = Object.assign({}, req.body)

  // slack secret token must be provided.
  if (!req.body || req.body.token !== process.env.SLACK_TOKEN) {
    return res.sendStatus(401)
  }

  // provides a respond function that any yargs
  // command can use to respond to Slack.
  context.respond = buildResponder(req.body.response_url)

  // run the yargs parser on the inbound slack command.
  parser.parse(req.body.text || '', context, (err, argv, output) => {
    if (err) logger.error(err.message)
    if (output) argv.respond(output)
  })

  res.send('')
})
```

`bodyParser` processes the inbound webhook from Slack and populates three
variables that are important to us:

* **req.body.token:** the shared secret between our application and
  and Slack. This allows us to verify that the post originated from Slack.
  _Note: make sure that you keep this token a secret! don't put it under
   version control._
* **req.body.text:** the text that the user typed into the Slack channel.
* **req.body.response_url:** the URL that your bot's response will be posted to.

### Responding to Slack

Slack provides a `response_url` in their webhook for a bot to post their response to.
Before passing `req.body.text` to yargs for processing, we populate a helper function that
allows yargs commands to easily send messages back to Slack:

```js
// returns a helper function for dispatching messages
// back to Slack.
function buildResponder (responseUrl) {
  return function (msg) {
    request.post({
      url: responseUrl,
      json: true,
      body: {
        response_type: 'in_channel',
        text: msg
      }
    }, function (err, res, body) {
      if (err) return logger.error(err)
      if (res && res.statusCode >= 400) logger.error('invalid response =', res.statusCode)
      else logger.info(body)
    })
  }
}
```

### Parsing the message with yargs

Creating a yargs instance for processing chat messages is almost identical to how you
would configure it for a command line application.

We begin by configuring the parser in the abstract:

```js
const parser = require('yargs')
  .usage('/joe [command]')
  .commandDir('commands')
  .demand(1)
  .strict()
  .help()
  .epilog("yargs' slack-bot Pirate Joe")
```

[`yargs.commandDir('commands')`](https://github.com/yargs/yargs#commanddirdirectory-opts) indicates that
we should load all the chat commands located in `/commands`.

We create a command JavaScript file for each of our chat commands.

Here's the module for translating English strings to Pirate strings:

**pirate.js**

```js
const pirateSpeak = require('pirate-speak')

exports.command = 'pirate <strings...>'
exports.describe = 'US English to US Pirate translator'
exports.builder = {}
exports.handler = function (argv) {
  argv.respond(pirateSpeak.translate(argv.strings.join(' ')))
}
```

With the yargs instance configured, we now simply run [`parse()`](https://github.com/yargs/yargs#parseargs-context-parsecallback) on each
of the inbound messages from Slack:

```js
context.respond = buildResponder(req.body.response_url)
parser.parse(req.body.text || '', context, (err, argv, output) => {
  if (err) logger.error(err.message)
  if (output) argv.respond(output)
})
```

Again, by providing the `context.respond` method, we give yargs' commands a
mechanism for sending messages back to Slack.

If any default output has been logged by yargs (perhaps `help` was executed), the
`parse()` function itself dispatches the `output` back to Slack.

### Deploying Your Application

The easiest way to get your slack-bot up and running is to create a Heroku application.
A [wonderful interactive tutorial](https://devcenter.heroku.com/articles/getting-started-with-nodejs#introduction) is
available on this topic.

Once your application is in the wild, visit:

`https://[your-slack].slack.com/apps/manage/custom-integrations`

and configure a new [Slash Command](https://api.slack.com/slash-commands) for your application.

## License

ISC

[travis-url]: https://travis-ci.org/yargs/pirate-joe
[travis-image]: https://img.shields.io/travis/yargs/pirate-joe/master.svg
[coveralls-url]: https://coveralls.io/github/yargs/pirate-joe
[coveralls-image]: https://img.shields.io/coveralls/yargs/pirate-joe.svg
[standard-image]: https://img.shields.io/badge/code%20style-standard-brightgreen.svg
[standard-url]: http://standardjs.com/
[standard-version-image]: https://img.shields.io/badge/release-standard%20version-brightgreen.svg
[standard-version-url]: https://github.com/conventional-changelog/standard-version


CHANGELOG.md
# Change Log

All notable changes to this project will be documented in this file. See [standard-version](https://github.com/conventional-changelog/standard-version) for commit guidelines.

<a name="1.0.0"></a>
# 1.0.0 (2016-10-15)


### Bug Fixes

* actually use the flipGuy, we need unit tests ([6e66e7c](https://github.com/yargs/pirate-joe/commit/6e66e7c))
* add the flip guy ([5009c67](https://github.com/yargs/pirate-joe/commit/5009c67))
* added engines field for heroku ([771d084](https://github.com/yargs/pirate-joe/commit/771d084))
* missing closing brace, yep we need tests ([0dd91e9](https://github.com/yargs/pirate-joe/commit/0dd91e9))
* slight tweaks to README; fixed bug with 200 response ([0d7c6e2](https://github.com/yargs/pirate-joe/commit/0d7c6e2))
* tiny tweaks to yargs output ([1e6062e](https://github.com/yargs/pirate-joe/commit/1e6062e))
* upgrade to version of yargs that fixes context issues ([6292a06](https://github.com/yargs/pirate-joe/commit/6292a06))
* we should not be calling flip on our pirate string ([bb820f3](https://github.com/yargs/pirate-joe/commit/bb820f3))
* whoops, we were referencing the wrong arg ([b272b5d](https://github.com/yargs/pirate-joe/commit/b272b5d))


### Features

* added pirate-speak command ([968bf67](https://github.com/yargs/pirate-joe/commit/968bf67))
* roughing out chat-bot codebase ([5dcf057](https://github.com/yargs/pirate-joe/commit/5dcf057))


test.js
/* global describe, context, it, before, after */

// nock lets us fake posting a command back
// to the originating slack channel.
const nock = require('nock')
const request = require('request')
const startServer = require('./server')

require('chai').should()

describe('Pirate Joe', () => {
  let server = null
  // all tests run against an actual running
  // instance of the server.
  before(function (cb) {
    startServer(function (_err, _server) {
      server = _server
      return cb()
    }, {
      logger: {info: function () {}},
      port: 9999
    })
  })

  context('unauthenticated', function () {
    it('responds with a 401 if SLACK_TOKEN is incorrect', function (done) {
      process.env.SLACK_TOKEN = 'token-batman'
      request.post({
        url: 'http://localhost:9999',
        form: {
          token: 'token-robin'
        }
      }, function (err, res, body) {
        if (err) return done(err)
        res.statusCode.should.equal(401)
        return done()
      })
    })
  })

  context('authenticated', function () {
    before(function () {
      process.env.SLACK_TOKEN = 'token-batman'
    })

    it('responds with help string, if help command given', function (done) {
      // response posted to slack.
      nock('https://response.example.com')
         .post('/commands/1234/5678')
         .reply(200, function (uri, requestBody) {
           requestBody.text.should.match(/US English to US Pirate translator/)
           return done()
         })

      // post help command.
      request.post({
        url: 'http://localhost:9999',
        form: {
          token: 'token-batman',
          text: 'help',
          response_url: 'https://response.example.com/commands/1234/5678'
        }
      }, function (err, res, body) {
        if (err) return done(err)
        res.statusCode.should.equal(200)
      })
    })

    describe('flip', function () {
      it('flips a string', function (done) {
        // response posted to slack.
        nock('https://response.example.com')
           .post('/commands/1234/5678')
           .reply(200, function (uri, requestBody) {
             requestBody.text.should.equal('(╯°□°）╯︵pʃɹoʍ oʃʃǝɥ')
             return done()
           })

        // post help command.
        request.post({
          url: 'http://localhost:9999',
          form: {
            token: 'token-batman',
            text: 'flip hello world',
            response_url: 'https://response.example.com/commands/1234/5678'
          }
        }, function (err, res, body) {
          if (err) return done(err)
          res.statusCode.should.equal(200)
        })
      })
    })

    describe('pirate', function () {
      it('translates english to pirate', function (done) {
        // response posted to slack.
        nock('https://response.example.com')
           .post('/commands/1234/5678')
           .reply(200, function (uri, requestBody) {
             requestBody.text.should.equal('ahoy how be ye today')
             return done()
           })

        // post help command.
        request.post({
          url: 'http://localhost:9999',
          form: {
            token: 'token-batman',
            text: 'pirate hello how are you today',
            response_url: 'https://response.example.com/commands/1234/5678'
          }
        }, function (err, res, body) {
          if (err) return done(err)
          res.statusCode.should.equal(200)
        })
      })
    })
  })

  after(function () {
    server.close()
  })
})


test/commands/name.js
/* global describe, it */

const name = require('../../commands/name')

require('chai').should()

describe('Name Command', () => {
  it('generates a random pirate name', (done) => {
    name.handler({
      user_name: 'bcoe',
      respond: (msg) => {
        msg.should.match(/bcoe your pirate name is/)
        return done()
      }
    })
  })
})


server.js
const bodyParser = require('body-parser')
const express = require('express')
const request = require('request')

// configure the yargs instance used
// to parse chat messages.
const parser = require('yargs')
  .usage('/joe [command]')
  .commandDir('commands')
  .demand(1)
  .strict()
  .help()
  .epilog("yargs' slack-bot Pirate Joe")

const app = express()
let logger = console

// Slack's slash commands are passed as an urlencoded
// HTTP post: https://api.slack.com/slash-commands
app.use(bodyParser.urlencoded({ extended: false }))

// slack webhook endpoint.
app.post('/', function (req, res) {
  let context = Object.assign({}, req.body)

  // slack secret token must be provided.
  if (!req.body || req.body.token !== process.env.SLACK_TOKEN) {
    return res.sendStatus(401)
  }

  // provides a respond function that any yargs
  // command can use to respond to Slack.
  context.respond = buildResponder(req.body.response_url)

  // run the yargs parser on the inbound slack command.
  parser.parse(req.body.text || '', context, (err, argv, output) => {
    if (err) logger.error(err.message)
    if (output) argv.respond(output)
  })

  res.send('')
})

// returns a helper function for dispatching messages
// back to Slack.
function buildResponder (responseUrl) {
  return function (msg) {
    request.post({
      url: responseUrl,
      json: true,
      body: {
        response_type: 'in_channel',
        text: msg
      }
    }, function (err, res, body) {
      if (err) return logger.error(err)
      if (res && res.statusCode >= 400) logger.error('invalid response =', res.statusCode)
      else logger.info(body)
    })
  }
}

// ping endpoint.
app.get('/', function (req, res) {
  res.send('pong')
})

// start our slack webhook server.
module.exports = function (cb, opts) {
  let port = process.env.PORT || 3000
  cb = cb || function () {}

  opts = opts || {}
  if (opts.logger) logger = opts.logger
  if (opts.port) port = opts.port

  const server = app.listen(port, function (foo) {
    if (process.env.HEROKU) keepAlive()
    logger.info('Pirate Joe bot listening on :' + port, 'beep boop')
    return cb(null, server)
  })
}

// ping our application every 5 minutes so
// that Heroku does not sleep it.
function keepAlive () {
  setInterval(function () {
    logger.info('pinged', process.env.APP_URL)
    request.get(process.env.APP_URL)
  }, 300000)
}
